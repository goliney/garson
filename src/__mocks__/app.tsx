import React from 'react';
import { render } from 'ink-testing-library';
import { Text } from 'ink';

const initialView = <Text>Initializing tests...</Text>;

export const app = render(initialView);

app.unmount = jest.fn().mockImplementation(() => app.rerender(initialView));
