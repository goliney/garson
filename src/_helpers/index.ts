// Components
export { Question } from './components/question';
export { Line } from './components/line';

// Hooks
export { useKeyHandler } from './use-key-handler';
export { useEnterKeyHandler } from './use-enter-key-handler';
