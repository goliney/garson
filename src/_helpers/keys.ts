export const ARROW_UP = '\u001B[A';
export const ARROW_DOWN = '\u001B[B';
export const ARROW_LEFT = '\u001B[D';
export const ARROW_RIGHT = '\u001B[C';
export const ENTER = '\r';
export const CTRL_C = '\x03';
export const BACKSPACE = '\x08';
export const DELETE = '\x7F';
export const SPACE = ' ';
