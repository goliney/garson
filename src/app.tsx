import { Box, render } from 'ink';
import React from 'react';

export const app = render(<Box>Initializing...</Box>, {
  debug: false,
});
