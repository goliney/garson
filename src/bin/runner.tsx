import React from 'react';
import { Box, render } from 'ink';
import { GarsonConfig } from '../types';

const { rerender, unmount } = render(<Box>Initializing...</Box>, {
  debug: false,
});

export function runner({ results, prompts, action }: GarsonConfig) {
  const [{ prompt, key }, ...rest] = prompts;
  rerender(
    prompt(result => {
      rerender(<React.Fragment />); // clear the previous prompt from the screen
      const newResults = { ...results, [key]: result };
      if (rest.length === 0) {
        // it was the last prompt in a chain
        const actionResponse = action(newResults);
        if (actionResponse) {
          runner(actionResponse);
        } else {
          unmount();
        }
      } else {
        // there is another prompt
        runner({ results: newResults, prompts: rest, action });
      }
    }),
  );
}
