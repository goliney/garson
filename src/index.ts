import { KeyType, PromptCallback, Results, ActionCallback, GarsonHandlerParams } from './types';

export function garson(results?: Results) {
  return garsonHandler({ results });
}

function garsonHandler({ results = {}, prompts = [] }: GarsonHandlerParams) {
  return {
    prompt: (key: KeyType, prompt: PromptCallback) =>
      garsonHandler({ results, prompts: [...prompts, { key, prompt }] }),
    action: (action: ActionCallback) => ({
      results,
      prompts,
      action,
    }),
  };
}
