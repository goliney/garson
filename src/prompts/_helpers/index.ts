// Components
export { Question } from './components/question';

// Misc
export { useKeyHandler } from './use-key-handler';
