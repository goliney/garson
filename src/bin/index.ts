import { runner } from './runner';

const cwd = process.cwd();
let garsonConfig;

try {
  // eslint-disable-next-line import/no-dynamic-require, global-require
  garsonConfig = require(`${cwd}/garson.config.js`);
} catch (e) {
  console.error('Failed to load config', e);
}

if (garsonConfig) {
  runner(garsonConfig);
}
