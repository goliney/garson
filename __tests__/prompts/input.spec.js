import { garson, prompts } from '../../src';
import { runner } from '../../src/bin/runner';
import { app } from '../../src/app';
import { ENTER } from '../../src/_helpers/keys';
import { write, trimFrame, stripColorsFromLastFrame } from '../helpers';

jest.mock('../../src/app');

const actionSpy = jest.fn();

const config = garson()
  // first prompt
  .prompt(
    'firstName',
    prompts.input({
      message: "What's your first name?",
      placeholder: 'E.g. John',
    })
  )
  // second prompt
  .prompt(
    'lastName',
    prompts.input({
      message: "What's your last name?",
      placeholder: 'E.g. Smith',
      defaultValue: 'Bing',
    })
  )
  // final action
  .action(actionSpy);

const configNoLabels = garson()
  .prompt('noLabels', prompts.input())
  .action(actionSpy);

describe('Input', () => {
  beforeEach(() => {
    actionSpy.mockClear();
  });

  afterEach(() => {
    app.unmount();
  });

  test('Normal flow', async () => {
    runner(config);
    await new Promise(resolve => setTimeout(resolve));
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write('Chandler');
    expect(trimFrame(stripColorsFromLastFrame())).toMatchSnapshot();

    await write(ENTER);
    expect(trimFrame(stripColorsFromLastFrame())).toMatchSnapshot();

    await write(ENTER);
    expect(actionSpy).toHaveBeenCalledWith({ firstName: 'Chandler', lastName: 'Bing' });
  });

  test('No labels', async () => {
    runner(configNoLabels);
    await new Promise(resolve => setTimeout(resolve));
    expect(trimFrame(stripColorsFromLastFrame())).toMatchSnapshot();

    await write('Something');
    expect(trimFrame(stripColorsFromLastFrame())).toMatchSnapshot();

    await write(ENTER);
    expect(actionSpy).toHaveBeenCalledWith({ noLabels: 'Something' });
  });
});
