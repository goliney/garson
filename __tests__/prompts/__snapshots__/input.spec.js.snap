// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input No labels 1`] = `"?"`;

exports[`Input No labels 2`] = `"? Something"`;

exports[`Input Normal flow 1`] = `"? What's your first name? E.g. John"`;

exports[`Input Normal flow 2`] = `"? What's your first name? Chandler"`;

exports[`Input Normal flow 3`] = `"? What's your last name? Bing"`;
